WordPress/src/main/java/org/wordpress/android/ui/reader/discover/interests/ReaderInterestsViewModel.kt

```diff
@@ -1,57 +1,151 @@
 ) : ViewModel() {
     var initialized: Boolean = false
 
-    private val _uiState = MutableLiveData<UiState>()
+    private val _uiState: MutableLiveData<UiState> = MutableLiveData(
+        UiState(listOf(), ReaderTagList(), DoneButtonHiddenUiState)
+    )
     val uiState: LiveData<UiState> = _uiState
 
+    private val _navigateToDiscover = MutableLiveData<Event<Unit>>()
+    val navigateToDiscover: LiveData<Event<Unit>> = _navigateToDiscover
+
+    private val selectedInterests = ReaderTagList()
+
     fun start() {
-        if (initialized) return
+        if (initialized) {
+            val uiState = uiState.value as UiState
+            updateUiState(uiState.copy(interestsUiState = getInterestsUiStateWithSelectedStates()))
+            return
+        }
         loadInterests()
     }
 
     private fun loadInterests() {
         viewModelScope.launch {
             val tagList = readerTagRepository.getInterests()
             if (tagList.isNotEmpty()) {
-                updateUiState(UiState(transformToInterestsUiState(tagList), tagList))
+                updateUiState(
+                    UiState(
+                        transformToInterestsUiState(tagList),
+                        tagList,
+                        getDoneButtonUiState()
+                    )
+                )
                 if (!initialized) {
                     initialized = true
                 }
             }
         }
     }
 
+    fun onInterestAtIndexToggled(index: Int, isChecked: Boolean) {
+        uiState.value?.let {
+            val enableDoneButton = selectedInterests.isEmpty() && isChecked
+            val disableDoneButton = selectedInterests.size == 1 && !isChecked
+
+            updateSelectedInterests(it.interests[index])
+
+            if (enableDoneButton || disableDoneButton) {
+                val uiState = uiState.value as UiState
+                updateUiState(
+                    uiState.copy(
+                        interestsUiState = getInterestsUiStateWithSelectedStates(),
+                        doneButtonUiState = getDoneButtonUiState()
+                    )
+                )
+            }
+        }
+    }
+
+    fun onDoneButtonClick() {
+        viewModelScope.launch {
+            readerTagRepository.saveInterests(selectedInterests)
+            _navigateToDiscover.value = Event(Unit)
+        }
+    }
+
+    private fun updateSelectedInterests(interestAtIndex: ReaderTag?) {
+        if (!selectedInterests.contains(interestAtIndex)) {
+            selectedInterests.add(interestAtIndex)
+        } else {
+            selectedInterests.remove(interestAtIndex)
+        }
+    }
+
     private fun transformToInterestsUiState(interests: ReaderTagList) =
         interests.map { interest ->
-            InterestUiState(
-                interest.tagTitle
-            )
+            InterestUiState(interest.tagTitle)
+        }
+
+    private fun getInterestsUiStateWithSelectedStates(): List<InterestUiState> {
```

Comment:
Do we needs this method? Do we even need the `selectedInterests` field? Can't we simply store the check/notChecked state inside the uiState and consider the uiState as the single source of truth?

Reply:
Refactored here: bfaaee9